large refactoring

kalmarek opened the pull request:
@DanielBrosch I hope this is not too scary!
You should still be able to recognize the code you wrote, but organized a bit and wrapped up in some layers of abstraction.

e.g. in `src/eigen_decomposition.jl` I've created `struct Eigenspace` and `struct EigenDecomposition` to make writing the Murota algorithm easier.
Once we compute `ED = EigenDecomposition(...)` and we have a randomized `Q'AQ` the whole inner loop looks like this:
```julia
    # compute the equivalence relation K (eq. 4.2)
    k = length(ED)
    K = IntDisjointSets(k) # tracks the merging of eigenspaces
    for i in 1:k
        for j in (i+1):k
            in_same_set(K, i, j) && continue # never happens!
            Ei, Ej = ED[i], ED[j]
            dim(Ei) != dim(Ej) && continue
            if any(x -> abs(x) ≥ atol, Q′AQ[Ei, Ej])
                # the block must be an isomorphism, merge!
                union!(K, i, j)
            end
        end
    end
```

# File structure
I decided to break the large main file to a bunch of smaller ones. Here's the output of `tree` with a few comments
```
/
├── CITATION.bib
├── docs # no changes here
├── LICENSE
├── Manifest.toml
├── Project.toml
├── README.md
├── src
│   ├── compat.jl # definitions to keep backward compatibility
│   ├── diagonalize.jl # compute Q_hat (from Murota) by linking eigen-decomposition and irreducible decomposition + transformation of constraints
│   ├── eigen_decomposition.jl Eigen-decompositions + Murota: Algorithm 4.1 (simple components) + Remark 4.1 (irreducible components)
│   ├── partitions.jl # the Partition type + admissible_subspace
│   ├── SDPSymmetryReduction.jl # imports, includes and exports only
│   └── utils.jl # small utility functions
└── test
    ├── example812.dat
    ├── lovasz.jl # tests for Lovász-θ' on ER graphs q = 3,5,7
    ├── qapdata
    │   └── esc16j.dat
    ├── qap.jl # test for esc16j qap problem
    ├── runtests.jl
    └── sd_problems.jl # SD problems for Lovász and qap
```
# Major changes:
## Tests
Tests depend now on `JuMP`, `CSDP` and `MutableArithmetic`. They also run ~3× longer (10 → 30s here) from scratch, but tests themselves take less than `1.0s` to execute.

## `Partition`
1. `Partition` is parametrized by `eltype(P.P)`. By default `Partition(M)` creates `Partition{UInt16}` to store at most `65535` distinct values which should be more than enough for combinatorial applications. Since `Partition` is the starting point of most of the functionality, it is easy for the user to specify `P = Partition{UInt32}(M)` if desired.
2. @sebastiendesignolle I removed the shortcut you created for `Partition{<:Integer}` as it was clashing with the correctness of the constructor. If this is badly needed from your side we can discuss how to bring it back.
3. `coarsestPart(P,Q)` was renamed to `refine(P,Q)`.
4. `rndPart(P)` is now `randomize(Float64, P)` which calls internally `fill!(M, P, values=rand(Float64, P.n+1)` to allow something along `getRandMat`. I think that `fill!` corresponds better to the semantic of what we're doing here.
5. `admPartSubspace(C, A, b, verbose=false)` calls now to `admissible_subspace(C, A, b, verbose=verbose, rtol=...)`
As can be seen from performance graphs below I didn't change much here, only read the @DanielBrosch paper and Permenter thesis to align notions with the notation, comments and my understanding (which might be still lacking ;)
6. `unSymmetrize` → `desymmetrize` but essentially any more descriptive name would be better :D. @DanielBrosch do you have a reference for this algorithm?

### Proposal for further changes
I was thinking about: defining
 * `size(P::Partition, args...) = size(P.P, args...)` and
 * `dim(P::Partition) = P.n`

to abstract from the internal representation of `P`. I think that further performance advances can be enabled here by some fancy partition refinement algorithms which I know nothing about ;).

## `EigenDecomposition` and `EigenSpace`
I was too impatient to understand the initial implementation (the abundance of indices made me feel very uncomfortable, as these are precisely moments where I can't be trusted :D) of Murota algorithm. So I decided to put some work and abstract what is needed.
It's a from-scratch implementation, but in the end not that much different, as we both follow the paper rather closely ;) Here are a few points to keep in mind:

1. In `eigen_decomposition(P::Partition, ...)` implementing Algorithm 4.1 We throw our own `InvalidDecompositionFiled` exception when decomposition field is incompatible with `T` provided by the user. I'm still torn between throwing exceptions and returning descriptive statuses. Let me know what do you think.
2. We use `K::DataStructures.IntDisjointSets` to maintain the merging of eigenspaces. **This adds `DataStructures.jl` as dependency**
3. `irreducible_decomposition` follows mostly the old code, the only difference is that the final permutation of columns is computed and applied in-place as an actual permutation.

### Proposal for further changes
Inspecting the snippet above of merging of eigenspaces:
 1. `in_same_set(K,i, j)` will never return true (or does it?!)
 2. We don't need to compute the whole of `Q'AQ` since we only access upper-triangular block of it (2× speedup possible)
 3. We could compute `Q′AQ` lazily and only realize it during the iteration in `if any(x -> abs(x) ≥ atol, Q′AQ[Ei, Ej])` condition. This should be fairly simple to achieve but probably of limited utility as almost all of those blocks will be 0, so short-circuiting will happen only rarely.
 4. ~There's a `TODO` at the end of `irreducible_decomposition`: given that we are interested in the first bunch of columns of `Base.permutecols!!(QKi * Pi, perm)` we could compute only those! I believe this will speed things up quite a bit as `n_merged_blks` is small compared to `size(Pi,2)`. Probably number one to investigate!~ Implemented in [`b8edc45`](https://github.com/DanielBrosch/SDPSymmetryReduction.jl/pull/4/commits/b8edc451d737a515931c7e9814e302d93153b98a)

## Diagonalization routines
I decided to decouple computing the diagonalization map (i.e. computing of `Q_hat`) and applying it to constraints (computing block forms of `P.P .== i`). In my experience in integrating `SymbolicWedderburn.jl` the isomorphism is more useful than the actual images and it's easy to chain them together. Furthermore for integrators it will be easier to compute the diagonalization map, check if it satisfies the consistency conditions over given `T` and than take decisions.

Also: The computation of the image of our basis
```julia
   [[Qi'*(P.P .== i)*Qi for Qi in Q_hat] for i in 1:P.n]
```
can be efficiently parallelized over `i` (see the current implementation)   and it should be up to the user whether to use BLAS threading (computation for each `i` is substantial), or set BLAS number of threads to `1` and use batching (e.g. when `P.n` is very large and hence `P.P .== i` are very sparse).

I have been makeing those consideration in `SymbolicWedderburn.jl` and I came up with something like [this](https://github.com/kalmarek/SymbolicWedderburn.jl/blob/d9266a71ea42b4d623c75dd71c0666db752f38ab/src/wedderburn_decomposition.jl#L114)
which is questionable at best ;) But for my applications constraints come in very strange [forms and formats](https://github.com/kalmarek/PropertyT.jl/blob/a10132e441d333a7ec78a3817771d792f98216cd/src/constraint_matrix.jl#L25) so this is the fastest thing I came up with. (the implementation of `_constraints(P)` and `sparse_constraint!` I proposed here is heavily inspired by this).

# Performance
I've run a benchmark on ER-graphs for primes 5-59 (benchmark code below). For `q=59` we have `P.n == 96`, `size(P,1) == 3541` and
| commit                 | Admissible subspace(s) | Block diagonalization (s) |
| ---------------------- | ---------------------- | ------------------------- |
| `d32c9b` (HEAD of #3)  | 14.2                   | 74.5                      |
| `6ad1d7`(HEAD of here) | 9.1                    | 15.8                      |

While this may look impressive, this

![timings](https://github.com/user-attachments/assets/3a7eb3c6-ab6d-4b5e-be43-fb7c95e3a431)

shows that I've just cut the constants ;) It was a hell enjoyable pruning though!

## Benchmark code
```julia
interesting(t) = (time=t.time, bytes=t.bytes, allocs=t.gcstats.allocd)

function bench(q)
    PG2q = vcat([[0, 0, 1]],
        [[0, 1, b] for b = 0:q-1],
        [[1, a, b] for a = 0:q-1 for b = 0:q-1])
    Adj = [x' * y % q == 0 && x != y for x in PG2q, y in PG2q]

    N = length(PG2q)
    C = ones(N^2)
    A = vcat(vec(Adj)', vec(Matrix{Float64}(I, N, N))')
    b = [0.0, 1.0]

    t_adms = @timed P = SR.admPartSubspace(C, A, b, false)
    t_blkd = @timed _ = SR.blockDiagonalize(P, false)

    @info (q, P.n, size(P.P, 1), t_adms.time, t_blkd.time)

    return (q=q, dim=P.n, n=size(P.P, 1), t_adms=interesting(t_adms), t_blkd=interesting(t_blkd))
end
```

DanielBrosch commented:
Thank you very much! Give me a bit of time to go through it, will try to get around to it soon.

Did you also use [Jet.jl](https://github.com/aviatesk/JET.jl) to check for additional potential issues? It usually seems to catch more than @code_warntype.

kalmarek commented:
@DanielBrosch Indeed Jet seems rather happy with the code ;)
I found one not dangling variable, and other than this on `admissible_subspace`, `diagonalize` and `basis_image`:
* there're some errors related to `@info` and `@timed`
* there's one error hidden deep inside QR fractorisation

on `blockDiagonalize` Jet reports some runtime dispatch and failure to optimize due to recursion in `clamptol!`, but this is gone when calling `basis_image` directly. hmm...

I've never used the tool before, but seem to be ok to ignore.

kalmarek commented:
ok, so I went further and implemented the optimization in `irreducible_decomposition`:
* I noticed that we don't need a separate randomized `A` for each `Ki`, so I hoisted randomization out of the loop.
* we compute only the columns of `QKi * Pi` we need, i.e. the first one from each eigenspace
* we compute only the `Qi'·A·Qj` blocks we actually need (i.e. for `j ∈ Ki) and not the whole `Q'·A·Q` matrix. 

results for the data of Lovász θ' (`q = 53`)
```julia
@btime irr = SR.irreducible_decomposition($(args)...)
```
```
aa0a440 # before
488.249 ms (1127 allocations: 194.85 MiB)

b8edc45 # after
211.868 ms (938 allocations: 99.27 MiB)
```
Time spent: 2h :D

kalmarek commented:
ok, I think this is the final update here:
* I decided to use [`Krylov.craig`](https://jso.dev/Krylov.jl/stable/solvers/ln/#Krylov.craig) instead of `A \ b` as it is ~2x faster for least-norm solutions
* rounding was taking quite some time because it does not vectorize; Since we don't care about fp correctness, but only that numbers close will be grouped together, I used shift-trunc-unshift approach which does vectorize. 
* I went in and implemented the `AbstractPartition` interface. There's a (failed -- performance wise) attempt at using bitsets to implement the interface. I added it to tests. We can decide not to include this change, it does look and feel nicer though, that is in my opinion ;) 

# Performance
I've run a benchmark on ER-graphs for primes 5-59 (benchmark code below). For `q=59` we have `P.n == 96`, `size(P,1) == 3541` and
| commit                 | Admissible subspace(s) | Block diagonalization (s) |
| ---------------------- | ---------------------- | ------------------------- |
| `d32c9b` (HEAD of #3)  | 14.2                   | 74.5                      |
| `6ad1d7`(prev HEAD)    | 9.1                    | 15.8                      |
| `01e7324` (curr HEAD)  | 7.0                    | 10.7                      |

Plots starting from the last update:
 
![timings_aa0a44_b8edc45_4e31f7_01e732](https://github.com/user-attachments/assets/8dc3cd3b-5cea-4c70-84bc-ddeb41bed760)

DanielBrosch commented:
@kalmarek Sorry for the delay! I started going trough the code this week, will have a couple comments/changes for you soon.

DanielBrosch commented:
I finally have some time to look at this. Thank you @kalmarek  (and @sebastiendesignolle !) for the hard work! This must have taken some time. It was on my to-do list for a while, but I doubt I could have improved it this much. I fixed some minor things, and have a few small comments:

## General
- To be honest, we should probably deprecate the old functions in compat.jl, and export the new functions which follow Julia style guides instead.
## Partitions
- UInt16 as default for Partition may be too small? It would certainly not be enough for some of my combinatorial applications (sadly). But I agree that in those cases one can just specify the type manually.
- Why are the functions `Base.fill!` so different for complex and real numbers? The complex variant always sets the zero entries to zero, the real one does not. The complex version also always returns a symmetric matrix.
- I would suggest only allowing the user to set the non-zero entries (i.e. `values` should have length `dim(P)`):
```julia
function Base.fill!(M::AbstractMatrix, P::Partition; values::AbstractVector)
    for idx in eachindex(P.matrix, M)
        P.matrix[idx] == 0 && continue
        M[idx] = values[P.matrix[idx]]
    end
    return M
end
```
Or do you have a use for setting zero entries to non-zeros?
- Maybe we can modify `refine!` to instead modify the first partition in place, following Julia Base. But we would either need a constructor for `Partition` which modifies the integer matrix in place, or split off the `unique -> integer` coefficient sorting. I added a simple implementation in `test/partitions_in_place.jl`. It appears to be about 10% faster than the current implementation for `q=59` (but I am running the benchmarks on a laptop, thus the benchmark times are quite variable). I had to use an int-pointer `Ref{Int}` for `nparts` to avoid using a mutable struct. Is there a better way? 
- I suspect partitions are generally too dense for `BitSet`s to be worth it, but thank you for trying!

## EigenDecomposition/Space and Diagonalization
- Throwing an exception when `T` is too small seems most natural here.
- I quite like the implementation of the Eigenspace decompositions! Makes it much nicer to use it.
- The decoupling of isomorphism and basis image makes a lot of sense. Quite often I just apply the symmetry reduction by rewriting things in terms of the image of the basis. But indeed, for polynomial optimization (and similar) it should be better to immediately compute the images of the matrices corresponding to, e.g., the monomials in the products. I will try and see how well this works for my `FlagSOS` package soon. (It is also in desperate need of a rewrite. I am currently in the process of splitting off multiple of the algorithms into separate packages, such as the Schreier-Sims algorithm (Your `Groups` package does not have it implemented, right?) and my Nauty implementation, and the decomposition algorithm for quotients of permutation modules. There is also an implementation of the regular-* representation somewhere in there, it would make a lot more sense here...)

## Others
- More of a Julia question: What is the point of let blocks like this:
```julia
tmp = let tmp = Vector{Matrix{T}}(undef, length(Q))
    for (i, q) in zip(eachindex(tmp), Q)
        tmp[i] = Matrix{T}(undef, size(P, 1), size(q, 2))
    end
    tmp
end
```
Isn't this exactly the same as the code without the let block?
- I am guessing that the intention may be to integrate this with  `JuMP`. Something like `@variable(m, X[1:n, 1:n], PSD, pattern)` would be quite useful. Not sure how complicated this would be. 

@kalmarek P.S.: I should take a closer look at your `SymbolicWedderburn` package, it sounds wonderful. Maybe we can have a chat about it some time soon?

DanielBrosch commented:
The assert `@assert first(Ki) == i` in `eigen_decomposition` just failed for me for a bigger problem. Sadly, I did not save the partition. Will try to break it again. Any idea what may have happened?

DanielBrosch commented:
Ah, we are not checking the block sizes in `diagonalize`. I assume Complex numbers are needed in my case. Any reason not to add it to `diagonalize`?

DanielBrosch commented:
I added an example of size 64x64 with $S_2$ symmetry in test/numerical_issues.jl. It causes the `@assert first(Ki) == i` every couple hundred calls.

kalmarek commented:
## Eigendecompositions
Great that you like the new design, I had some fun understanding and refactoring it!

- [ ] try to figure out what is happening in the example you provided when the assertion fails

### Murota performance
I have one more question for you. In the Murota algorithms we have (randomized) `A` which we need to transform into `Q'*A*Q` and later we need inspect ~half of it in blocks. We can either
1. compute the triple matrix product `Q′AQ = Q'*A*Q` and then take a view of the blocks: `Q′AQ[Ei, Ej]`, or
2. only compute the blocks on-demand: `block(A, Ei, Ej)`.

The former is faster for smaller sizes, but the latter (by avoiding some work) should be faster later on. Now, I don't expect Murota algorithm to be successfully applied used when `size(P,1) > 10_000` (I'm looking at you `eigen`), but optimizing for small sizes where execution is quick anyway is pointless ;) Can you
- [ ] check with larger examples which option is faster?

### Murota further
* The call to `inv` in `irreducible_decomposition` looked fishy to me and indeed I found two later papers of Murota et al where they replace it by transpose/adjoint, i.e. `block(A, Ei, Ej)` is already orthogonal as expected.
- [ ] replace `inv`s with `'` in `irreducible_decomposition`

* These papers develop the fully general version for complex and quaternionic algebras. I'd be happy to (help) implementing those, but only when we have an actual use for them 😃.

kalmarek commented:
## Others
* There's no point for keeping this particular `let` block, it does exactly the same as without the local scope; Sometimes though the compiler fails to infer properly the types and let blocks help it to do it.
  In general I use let blocks as logical separation of code into more digestible portions, which do not pollute the scope of functions (e.g. I'm done with processing `x` at this point).
* The integration with JuMP/MOI will probably happen at much higher level, but for this we have @blegat as a specialist ;) he would be also interested probably in the [FlagSOS](https://github.com/DanielBrosch/FlagSOS.jl/tree/main) project of yours.
* As for an implementation of Schreier-Sims here's one for permutation groups: https://github.com/kalmarek/PermutationGroups.jl/blob/master/src/schreier_sims.jl 
  However if you're more into `AbstractAlgebra.jl`/`Nemo.jl`/`Oscar.jl` world I'd stick with what they use. They dropped the support for `GroupsCore.jl` some time in the past, which means none of my code for groups will work with their objects.
* I'll have a look at the regular *-representation code you have. Is it [this one](https://link.springer.com/content/pdf/10.1007/s10107-011-0461-3.pdf)?
* For the meeting, write me an email, we can plan something early next week!

kalmarek commented:
@DanielBrosch I don't have a windows machine to test this, but this should work;

how do you feel about merging this?

sebastiendesignolle commented:
Great to see that this is merged, thanks!

DanielBrosch commented:
Sorry it took so long! Was fighting with some deadlines (and am very good at procrastinating). I'll try to finish the new version asap, for now you can use 
`pkg> add SDPSymmetryReduction#main`
(I think).
